# **Business Understanding**

**Enterprise LLM Usage, Performance & User Satisfaction Dataset**

**Overview**

This dataset is a synthetic Generative AI (GenAI) and Large Language Model (LLM) usage dataset designed for data science, machine learning, and analytics projects.

It simulates real-world interactions between users and enterprise LLM applications by capturing information related to model usage, prompt characteristics, response performance, latency, token consumption, hallucinations, costs, and user satisfaction.

**Dataset Features**
Feature	    ->    Description
session_id		    ->    Unique identifier for each interaction
model_name		    ->    Name of the LLM used
application_domain		    ->    Industry or business domain
task_type		    ->    Type of AI task performed
prompt_length		    ->    Length of the user prompt
total_tokens	    ->    Total number of processed tokens
temperature	Sampling	    ->     temperature
top_p	    ->    	Top-p sampling parameter
rag_enabled	Whether	    ->     Retrieval-Augmented Generation was enabled
latency_sec	Model	    ->     response latency (seconds)
hallucination_flag	    ->    	Indicates hallucinated responses
user_satisfaction	    ->    	User satisfaction rating (1–5)
estimated_cost_usd	    ->    	Estimated API inference cost
successful_response	Binary	    ->     target indicating successful response

### **Loading Libraries**

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

### **Loading data**

In [4]:
df = pd.read_csv('data\\genai_llm_usage_dataset_1000.csv')

# **Data Understanding**

In [5]:
df.shape

(1000, 14)

The dataset containing one thousand observations


In [6]:
df.head()

,session_id,model_name,application_domain,task_type,prompt_length,total_tokens,temperature,top_p,rag_enabled,latency_sec,hallucination_flag,user_satisfaction,estimated_cost_usd,successful_response
0,1,GPT-4o,Healthcare,Summarization,1319,2455,0.17,0.87,0,3.76,0,4,0.0236,1
1,2,GPT-4o,Customer Support,Summarization,486,1425,0.76,0.51,0,2.36,1,5,0.0236,0
2,3,Gemma 3,Retail,Classification,1439,1648,0.63,0.64,0,3.68,1,4,0.0139,0
3,4,Gemma 3,Retail,Summarization,1662,2812,0.07,0.73,0,3.25,1,4,0.0374,0
4,5,Gemma 3,Finance,Classification,1452,1678,0.10,0.83,0,8.04,0,4,0.0171,0


In [7]:
df.dtypes

session_id               int64
model_name                 str
application_domain         str
task_type                  str
prompt_length            int64
total_tokens             int64
temperature            float64
top_p                  float64
rag_enabled              int64
latency_sec            float64
hallucination_flag       int64
user_satisfaction        int64
estimated_cost_usd     float64
successful_response      int64
dtype: object

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   session_id           1000 non-null   int64  
 1   model_name           1000 non-null   str    
 2   application_domain   1000 non-null   str    
 3   task_type            1000 non-null   str    
 4   prompt_length        1000 non-null   int64  
 5   total_tokens         1000 non-null   int64  
 6   temperature          1000 non-null   float64
 7   top_p                1000 non-null   float64
 8   rag_enabled          1000 non-null   int64  
 9   latency_sec          1000 non-null   float64
 10  hallucination_flag   1000 non-null   int64  
 11  user_satisfaction    1000 non-null   int64  
 12  estimated_cost_usd   1000 non-null   float64
 13  successful_response  1000 non-null   int64  
dtypes: float64(4), int64(7), str(3)
memory usage: 109.5 KB


The dataset contains no missing values.

In [8]:
df.describe()

,session_id,prompt_length,total_tokens,temperature,top_p,rag_enabled,latency_sec,hallucination_flag,user_satisfaction,estimated_cost_usd,successful_response
count,1000.000000,1000.000000,1000.000000,1000.00000,1000.00000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,500.500000,1022.780000,1636.763000,0.74728,0.74207,0.502000,5.368310,0.507000,4.345000,0.018026,0.202000
std,288.819436,574.062391,681.390085,0.42545,0.14234,0.500246,2.375312,0.500201,0.581647,0.011971,0.401693
min,1.000000,12.000000,71.000000,0.00000,0.50000,0.000000,0.310000,0.000000,3.000000,0.000500,0.000000
25%,250.750000,520.750000,1118.000000,0.37000,0.62000,0.000000,3.300000,0.000000,4.000000,0.008400,0.000000
50%,500.500000,1056.000000,1635.500000,0.74000,0.74000,1.000000,5.460000,1.000000,4.000000,0.015400,0.000000
75%,750.250000,1493.250000,2142.250000,1.11000,0.86000,1.000000,7.370000,1.000000,5.000000,0.024900,0.000000
max,1000.000000,2000.000000,3158.000000,1.50000,1.00000,1.000000,10.440000,1.000000,5.000000,0.060300,1.000000


**Numerical Variable Analysis**

In [11]:
df['prompt_length'].describe()

count    1000.000000
mean     1022.780000
std       574.062391
min        12.000000
25%       520.750000
50%      1056.000000
75%      1493.250000
max      2000.000000
Name: prompt_length, dtype: float64

Insights:
- On average the prompts are 1056 characters long
- The longest prompt has 2000 characters
- The shortest prompt has 12 characters
- 25% of the prompts had less than 520 characters
- Half of the prompts had less than 1056 characters
- 75% of the prompts had less than 1493 characters

In [12]:
df['total_tokens'].describe()

count    1000.000000
mean     1636.763000
std       681.390085
min        71.000000
25%      1118.000000
50%      1635.500000
75%      2142.250000
max      3158.000000
Name: total_tokens, dtype: float64

Insights:
- On average each prompt used up 1636 tokens
- The most amount of tokens used is 3158
- The least amount of tokens used is 71
- 25% of the prompts contain fewer than 1118 tokens.
- Half of the prompts contain fewer than 1635 tokens.
- 75% of the prompts contain fewer than 2142 tokens.

In [ ]:
df[['latency_sec', "estimated_cost_usd"]].describe()

,latency_sec,estimated_cost_usd
count,1000.000000,1000.000000
mean,5.368310,0.018026
std,2.375312,0.011971
min,0.310000,0.000500
25%,3.300000,0.008400
50%,5.460000,0.015400
75%,7.370000,0.024900
max,10.440000,0.060300


Latency Insights:
- On average, an AI request takes approximately 5.368 seconds to complete.
- The longest AI request took 10.44 seconds
- The shortest AI request took 0.31 seconds
- 50% of the AI request took under 5.46 seconds
- 75% of the AI request took under 7.37 seconds

In [18]:
df[['prompt_length',"estimated_cost_usd"]].describe()

,prompt_length,estimated_cost_usd
count,1000.000000,1000.000000
mean,1022.780000,0.018026
std,574.062391,0.011971
min,12.000000,0.000500
25%,520.750000,0.008400
50%,1056.000000,0.015400
75%,1493.250000,0.024900
max,2000.000000,0.060300


Bivariate Analysis Insisght:
- prompt lenght influences the cost of the model (The longer the prompt the more it costs)
- Cost varies proportionally more than prompt length, suggesting that factors other than prompt length may also influence cost.

In [22]:
df[['total_tokens','prompt_length', "latency_sec","estimated_cost_usd"]].describe()

,total_tokens,prompt_length,latency_sec,estimated_cost_usd
count,1000.000000,1000.000000,1000.000000,1000.000000
mean,1636.763000,1022.780000,5.368310,0.018026
std,681.390085,574.062391,2.375312,0.011971
min,71.000000,12.000000,0.310000,0.000500
25%,1118.000000,520.750000,3.300000,0.008400
50%,1635.500000,1056.000000,5.460000,0.015400
75%,2142.250000,1493.250000,7.370000,0.024900
max,3158.000000,2000.000000,10.440000,0.060300


Multivariate Analysis Insisght:
- Prompt length is the primary driver of both cost and latency. Longer prompts require more processing which lead to higher latency and therefore costing more.

**Categorical Variable**

In [10]:
df['model_name'].describe()

count              1000
unique                6
top       Mistral Large
freq                193
Name: model_name, dtype: object

The data contains 6 different models

In [20]:
df["model_name"].value_counts()

model_name
Mistral Large    193
Qwen 2.5         177
GPT-4o           167
Gemma 3          164
Claude 3.7       151
Llama 3.1        148
Name: count, dtype: int64

Insights:
- Mistral Large is the most frequent category with 193 records.
- Llama 3.1 is the least frequent category with 148 records.

In [21]:
df["application_domain"].value_counts()

application_domain
Customer Support    165
Education           156
Retail              154
Finance             137
Healthcare          131
Coding              130
Legal               127
Name: count, dtype: int64

Insights:
- Customer Support is the most frequent category with 165 records.
- Legal is the least frequent category with 127 records.

# **Data Preparation**

# **Modeling**

# **Evaluation**

# **Deployment**